File: analyze_realworldtrf.m
close all; clear all; clc;
init_unfold;
dataFolder = './data';
dataFile = 'May20lsl_synchronize_test_02.xdf';

%% Load and preprocess the EEG
EEG = eeg_load_xdf(fullfile(dataFolder,dataFile));
EEG = pop_select(EEG,'nochannel',5); % Remove unused aux channel
EEG = pop_eegfiltnew(EEG, 0.1, 15);

%% Load the streams
streams = load_xdf(fullfile(dataFolder,dataFile));
eegStream = streams{1};
eegTimes = eegStream.time_stamps;
audioStream = streams{2};
audio = audioStream.time_series;
audio = rescale(audio,-1,1);
audioTimes = audioStream.time_stamps;
audioSR = str2num(audioStream.info.nominal_srate);
% plot(audioTimes,audio');

%% Envelope
htAudio = hilbert(audio);
envAudio = abs(htAudio);
fEnvAudio = lowpass(envAudio,128,audioSR);
fEnvAudio = mean(fEnvAudio,1);
fEnvAudio = rescale(fEnvAudio,0,1);
dsFEnvAudio = interp1(audioTimes,fEnvAudio,eegTimes);
% plot(eegTimes,dsFEnvAudio);

%% TRF
uEEG = EEG;
for j = 1:4
    uEEG.chanlocs(j).type = [];
end

% Add an event to EEG
uEEG.event(1).type = '1';
uEEG.event(1).latency = 1;
uEEG = eeg_checkset(uEEG);

% Flag bad segments of data
winrej = uf_continuousArtifactDetect(uEEG,'amplitudeThreshold',250,'windowsize',2000,'stepsize',1000,'combineSegments',[]);

% Time window for TRF
ufTime = [-0.5 1];

% Add audio signal to data
uEEG.data(5,:) = dsFEnvAudio;

% Generate design matrix
uEEG = uf_timeexpandDesignmat_addTRF(uEEG,'channel',5,'name','envelope','timelimits',ufTime);

% Remove audio channel and do artifact rejection
uEEG.data(5,:) = [];
uEEG = uf_continuousArtifactExclude(uEEG,struct('winrej',winrej));

% Grab the data and the design matrix
data = double(uEEG.data);
X = uEEG.unfold.Xdc;

% Flag zero rows
nonZero = any(X,2);
isZero = ~nonZero;
data(:,isZero') = [];
X(isZero',:) = [];

% Flag and remove any rows with NaN
isNan = isnan(sum(X,2));
data(:,isNan) = [];
X(isNan,:) = [];

%% Do cross-validation (optional)
% regtype = 'onediff';
% lambdas = [1E2 1E3 1E4 1E5 1E6 1E7 1E8 1E9 1E10];
% k = 10;
% [allErrors,bestBeta] = doRegCV(data,X,regtype,{1:size(X,2)},{[]},lambdas,k);
% figure();
% plot(allErrors);
% [~,j] = min(allErrors);
% disp(lambdas(j));

%% Solve GLM
regtype = 'onediff';
lambda = 1E5; % Get this from cross-validation above
thisPDM = pinv_reg(X,lambda,regtype);
tempBeta = thisPDM * data';

%%
figure();
plot(uEEG.unfold.times,tempBeta);
legend({uEEG.chanlocs.labels});

File: analyze_realworldtrf.py
# Script from Dr Hassall, rewritten in Python with the MNE-Python library
# Compute and return the TRF for a dataset with synchronized Audio and subsequently evoked EEG

import sys
import mne
import pyxdf
import numpy as np
import matplotlib.pyplot as plt

# Load data

def main():
    if not len(sys.argv) == 2:
        print("Invalid syntax. python analyze_realworld.trf.py [filename]")
        exit(0)
    dataFile = sys.argv[1]

    ## Load data to streams, header
    streams, header = pyxdf.load_xdf(dataFile)

    eegStreamNames = [str(ch["label"][0]) for ch in streams[0]["info"]["desc"][0]["channels"][0]["channel"]]
    audioStreamName = [streams[1]["info"]["name"][0]]

    eeg = streams[0]["time_series"].T
    eeg *= 1e-6 / 50 / 2
    eegSfreq = float(streams[0]["info"]["nominal_srate"][0])

    eegInfo = mne.create_info(eegStreamNames, eegSfreq, ["eeg", "eeg", "eeg", "eeg", "eeg"])
    eegRaw = mne.io.RawArray(eeg, eegInfo)
    eegRaw = eegRaw.drop_channels("Right AUX") # We are not using this "electrode"

    audio = streams[1]["time_series"].T
    audioSfreq = float(streams[1]["info"]["nominal_srate"][0])

    audioInfo = mne.create_info(audioStreamName, audioSfreq, ["misc"])
    audioRaw = mne.io.RawArray(audio, audioInfo)


    ## Preprocessing and envelope
    audioRaw.apply_function(lambda x:  mne.baseline.rescale(data=x, times=audioRaw.times, baseline=(-1,1)), picks="all")

    # Equalize length
    #print("HONK", eegRaw.first_time, audioRaw.first_time)

    htAudio = audioRaw.copy()
    htAudio.apply_hilbert(picks="all")

    envAudio = htAudio
    envAudio.apply_function(np.abs, dtype=np.dtype(np.float64), picks="all")

    fEnvAudio = envAudio
    fEnvAudio.filter(128, None, picks="all")
    fEnvAudio.apply_function(lambda x: np.mean(x, axis=0, keepdims=True), channel_wise=False, picks="all")
    fEnvAudio.apply_function(lambda x: mne.baseline.rescale(data=x, times=audioRaw.times, baseline=(0,1)), picks="all")

    ## We need to reshape the data (downsampled) so we need to create the RawArray from scratch.
    #dsFEnvAudio = np.interp(x=eegRaw.times, xp=audioRaw.times, fp=audioRaw.get_data()[0])
    dsFEnvAudio = fEnvAudio
    dsFEnvAudio = dsFEnvAudio.resample(eegSfreq, events=eegRaw.get_data(picks="all"))[0]


    ## TRF calculation
    uEEG = eegRaw

    uEEG.crop(tmax=dsFEnvAudio.times[-1])
    uEEG.add_channels([dsFEnvAudio])
    uEEG.filter(0.1, 15, picks="eeg")


    fEnvAudio.plot(scalings=None, duration=2, start=0, block=True)
    eegRaw.plot(scalings=dict(eeg=1e-6), duration=2, start=0, block=True)


if __name__ == "__main__":
    main()


# Load data
# Preprocessing
# Envelope settings
# Create Empty TRF
# Add event to EEG
# Flag bad segments of data
# Time window for TRF
# Add audio signal to data
# Generate design matrix w/ recorded data
# Perform artifact rejection on EEG
